src/nbl/video/IDescriptorPool.cpp

```diff
@@ -4,64 +4,166 @@
 namespace nbl::video
 {
 
-bool IDescriptorPool::createDescriptorSets(uint32_t count, const IGPUDescriptorSetLayout* const* layouts, core::smart_refctd_ptr<IGPUDescriptorSet>* output)
+uint32_t IDescriptorPool::createDescriptorSets(uint32_t count, const IGPUDescriptorSetLayout* const* layouts, core::smart_refctd_ptr<IGPUDescriptorSet>* output)
 {
-    core::vector<SDescriptorOffsets> descriptorOffsets(count);
+    core::vector<SDescriptorOffsets> descriptorOffsets;
+    descriptorOffsets.reserve(count);
+
+    system::ILogger* logger = nullptr;
+    {
+        auto debugCallback = getOriginDevice()->getPhysicalDevice()->getDebugCallback();
+        if (debugCallback)
+            logger = debugCallback->getLogger();
+    }
 
     for (uint32_t i = 0u; i < count; ++i)
     {
         if (!isCompatibleDevicewise(layouts[i]))
-            return false;
+        {
+            if (logger)
+                logger->log("Device-Incompatible descriptor set layout found at index %u. Sets for the layouts following this index will not be created.", system::ILogger::ELL_WARNING, i);
+
+            break;
+        }
         
-        descriptorOffsets[i] = allocateDescriptorOffsets(layouts[i]);
+        if (!allocateDescriptorOffsets(descriptorOffsets.emplace_back(), layouts[i]))
+        {
+            if (logger)
+                logger->log("Failed to allocate descriptor offsets in the pool's storage for descriptor set layout at index %u. Sets for the layouts following this index will not be created.", system::ILogger::ELL_WARNING, i);
+
+            descriptorOffsets.pop_back();
+            break;
+        }
+    }
+
+    const auto successCount = descriptorOffsets.size();
+    assert(count >= successCount);
+    std::fill_n(output + successCount, count - successCount, nullptr);
+
+    const uint32_t allocatedDSOffset = m_descriptorSetAllocator.alloc_addr(successCount, 1u);
+    if (allocatedDSOffset != m_descriptorSetAllocator.invalid_address)
+    {
+        if (createDescriptorSets_impl(successCount, layouts, descriptorOffsets.data(), allocatedDSOffset, output))
+        {
+            for (uint32_t i = 0u; i < successCount; ++i)
+                m_allocatedDescriptorSets[allocatedDSOffset + i] = output[i].get();
+
+            return successCount;
+        }
+    }
+    else
+    {
+        if (logger)
+            logger->log("Failed to allocate descriptor sets.", system::ILogger::ELL_ERROR);
     }
 
-    return createDescriptorSets_impl(count, layouts, descriptorOffsets.data(), output);
+    // Free the allocated offsets for all the successfully allocated descriptor sets.
+    for (uint32_t i = 0u; i < successCount; ++i)
+        freeDescriptorOffsets(descriptorOffsets[i], layouts[i]);
```

Comment:
you need a wholesale `cancel` for an entire `core::vector<SDescriptorOffsets>` range, its a "special free" because it will also be able to work when the DS Pool does not usually allow for freeing

Reply:
> because it will also be able to work when the DS Pool does not usually allow for freeing

[Oh, but it does](https://github.com/Devsh-Graphics-Programming/Nabla/blob/ba8162877f6f8d193a0cfb77463d037fa260624e/src/nbl/video/IDescriptorPool.cpp#L129). The only difference then would be that I write the `for` loop outside the cancel function rather than inside. The reason I do this is because then I can reuse it [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/ba8162877f6f8d193a0cfb77463d037fa260624e/src/nbl/video/IDescriptorPool.cpp#L108).

Reply:
```cpp
    for (uint32_t i = 0u; i < successCount; ++i)
        freeDescriptorOffsets(descriptorOffsets[i], layouts[i]);
 ```
 
 this calls `reset` on linear allocators from recent allocation to oldest, you don't end up resetting to the minimum value

Reply:
To avoid things like this https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files/1cd208bb98bdaa9f3eaef5530458ec23aeb2861b..ba8162877f6f8d193a0cfb77463d037fa260624e#r1121522223

if would be best to make the function
```cpp
void IDescriptorPool::freeDescriptorOffsets(const uint32_t count, const SDescriptorOffsets* beginOffsets, const IGPUDescriptorSetLayout* beginLayouts)
{
    for (uint32_t j = 0; j < count; j++)
        m_descriptorSetAllocator.free_addr(beginOffsets[j].dsOffset,1);

    for (uint32_t j = 0; j < count; j++)
    for (uint32_t i = 0; i <= static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT); ++i)
    {
        if (allowsFreeing())
             m_descriptorAllocators[i]->free(offsets.data[i], layout->getTotalDescriptorCount(static_cast<asset::IDescriptor::E_TYPE>(i)));
        else if (beginOffsets[j].data[i] != Invalid)
        {
             // first allocated offset will be the lowest
             m_descriptorAllocators[i]->linearAllocator.reset(offsets.data[i]);
             break;
        }
    }
}
```

Reply:
Resolved.